### PR TITLE
Randomize Global Program cron schedule minute

### PR DESCRIPTION
**Pull request by danfaizer**
This PR provides "randomization" in the minute defined by the cron schedule string for programs.

Two new configuration parameters have been added:
`RANDOMIZE_CRON_MINUTE_SUFFIXES` : Defines a list of suffixes which if matches the program ID, the cron schedule minute should be randomized.
`RANDOMIZE_CRON_MINUTE_INTERVAL`: Defines the interval for the randomized value. Values should be between 0 and 59.

The change works in the following way:
At crontinuous boot time, when reading to the existing scheduled entries, is evaluated if any of the RANDOMIZE_CRON_MINUTE_SUFFIXES matches the program ID, if so, the minute value for the cron string will be randomized "at job executor" level.

### CODE REVIEW

**Comment by ka3de** [comment]
Hey @danfaizer thanks for taking a look at this!

I have a general comment, not related with the implementation details. I have a concern that we are coupling too much the crontinuous implementation to the actual scans definition/logic etc. For example in [this function](https://github.com/adevinta/vulcan-crontinuous/blob/8f0b0ae7a16fba8bd5e0dbeb57033ec27808ae5c/crontinuous.go#L144-L158). In my opinion crontinuous should be pretty agnostic of whatever action is scheduling (even though we already have different entities for scan and reports).

Do you think we should make the randomization implementation apply for every scan schedule? So at least it's agnostic of scan types etc. In the end, a difference of 1 min for scan start time does not make much difference to end users even if we apply it for every scan, but it does make it for our system [unfortunately].

**Comment by danfaizer** [comment]
As discussed with @ka3de and @jesusfcr I've decoupled the program filter/selection from the API Global Programs and defined a more flexible way to select programs to randomize minutes.

In regards of minute interval calculation we've agreed to leave it as it is as it only has sense with Global Programs and it doesn't make any difference to have a more complex randomize interval formula.

In regards of applying the randomize to all program we agreed it won't be a desired effect as users can create their own programs which may require to be triggered at a specific time.

**Review by jesusfcr — CHANGES_REQUESTED** [review]
Just two changes

**Review by jesusfcr — DISMISSED** [review]
Nice solution!

File: local.env [hunk 1]
```diff
@@ -10,3 +10,5 @@ ENABLE_TEAMS_WHITELIST_SCAN=false
 TEAMS_WHITELIST_SCAN=[]
 ENABLE_TEAMS_WHITELIST_REPORT=false
 TEAMS_WHITELIST_REPORT=[]
+RANDOMIZE_CRON_MINUTE_SUFFIXES=""
```
Comment (jesusfcr) [inline local.env:13]: I would remove the ""

**Review by jesusfcr — DISMISSED** [review]

File: crontinuous_test.go [hunk 2]
```diff
@@ -332,6 +332,137 @@ func TestExecutesEntries(t *testing.T) {
...
+			name:            "Two prefixes, one matching",
+			programID:       "program@redcon-scan",
+			programSuffixes: "@redcon-scan,@periodic-full-scan",
+			want:            true,
+		},
+		{
+			name:            "Two prefixes, one matching",
+			programID:       "program@redcon-scan",
+			programSuffixes: "@redcon-scan,@periodic-full-scan",
+			want:            true,
```
Comment (manelmontilla) [inline crontinuous_test.go:370]: I think this test is duplicated it is the same than the test above, isn't it?

**Review by manelmontilla — DISMISSED** [review]
Just a couple of minor things that can be, indeed, addressed in another PR if you think it's worth it.

File: crontinuous.go [hunk 3]
```diff
@@ -87,6 +94,9 @@ func NewCrontinuous(cfg Config, logger *logrus.Logger,
 	scanCreator ScanCreator, scanCronStore ScanCronStore,
 	reportSender ReportSender, reportCronStore ReportCronStore) *Crontinuous {
 
+	if cfg.RandomizeCronMinuteInterval < 1 || cfg.RandomizeCronMinuteInterval > 59 {
```
[the remark this answers is not in the document]
Comment (danfaizer) [reply crontinuous.go]: Well, most common case will be that RandomizeCronMinuteInterval is set to `0` (which is the default value) and we want to set just by convenience `59`.
In any case, it have sense to add a log line. Added here [a732854](https://github.com/adevinta/vulcan-crontinuous/pull/19/commits/a732854aefc9370ebca0ced91b2b5db52a3658e1)

File: crontinuous_test.go (hunk 2 above)
Comment (danfaizer) [reply crontinuous_test.go]: Absolutely right. Removed here [a732854](https://github.com/adevinta/vulcan-crontinuous/pull/19/commits/a732854aefc9370ebca0ced91b2b5db52a3658e1)

**Review by manelmontilla — APPROVED** [review]
LGTM

**Pull request merged**